File: src/Tiedragon.XmppMessenger.Core/Xmpp/XmppMucAvatar.cs
namespace Tiedragon.XmppMessenger.Core.Xmpp;

public static class XmppMucAvatar
{
    public const string RoomInfoFormType = "http://jabber.org/protocol/muc#roominfo";

    public const string AvatarHashField = "muc#roominfo_avatarhash";

    public static bool SupportsAvatars(XmppServiceDiscoveryInfo info)
    {
        ArgumentNullException.ThrowIfNull(info);
        return info.Supports(XmppMultiUserChat.NamespaceName)
            && info.Supports(XmppVCardTemp.NamespaceName);
    }

    public static IReadOnlyList<string> GetRoomAvatarHashes(XmppServiceDiscoveryInfo info)
    {
        ArgumentNullException.ThrowIfNull(info);

        var form = info.DataForms.FirstOrDefault(form =>
            string.Equals(form.FormType, RoomInfoFormType, StringComparison.Ordinal));
        if (form is null || !form.Fields.TryGetValue(AvatarHashField, out var values))
        {
            return [];
        }

        return values
            .Select(value => value.Trim().ToLowerInvariant())
            .Where(value => value.Length > 0)
            .Distinct(StringComparer.Ordinal)
            .ToArray();
    }

    public static XmppIq CreateGetRequest(string id, XmppAddress room)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(id);
        ArgumentNullException.ThrowIfNull(room);

        return XmppVCardTemp.CreateGetRequest(id, XmppAddress.Parse(room.Bare));
    }

    public static XmppIq CreateSetRequest(
        string id,
        XmppAddress room,
        byte[] imageData,
        string contentType = XmppUserAvatar.RequiredContentType)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(id);
        ArgumentNullException.ThrowIfNull(room);
        ArgumentNullException.ThrowIfNull(imageData);
        if (imageData.Length == 0)
        {
            throw new ArgumentException("Room avatar image data cannot be empty.", nameof(imageData));
        }

        var vCard = new XmppVCardTemp(Photo: XmppVCardAvatar.CreatePhoto(imageData, contentType));
        return XmppVCardTemp.CreateSetRequest(id, vCard, XmppAddress.Parse(room.Bare));
    }

    public static XmppIq CreateRemoveRequest(string id, XmppAddress room)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(id);
        ArgumentNullException.ThrowIfNull(room);

        return XmppVCardTemp.CreateSetRequest(id, new XmppVCardTemp(), XmppAddress.Parse(room.Bare));
    }

    public static bool TryParseVerifiedAvatar(
        XmppIq iq,
        IReadOnlyList<string> advertisedHashes,
        out XmppMucAvatarPhoto? photo)
    {
        photo = null;
        ArgumentNullException.ThrowIfNull(advertisedHashes);

        if (!XmppVCardTemp.TryParseResult(iq, out var vCard)
            || vCard?.Photo?.Base64Data is null)
        {
            return false;
        }

        byte[] bytes;
        try
        {
            bytes = Convert.FromBase64String(vCard.Photo.Base64Data);
        }
        catch (FormatException)
        {
            return false;
        }

        var hash = XmppVCardAvatar.ComputePhotoHash(bytes);
        var matched = advertisedHashes.Count == 0
            || advertisedHashes.Any(value => string.Equals(value, hash, StringComparison.OrdinalIgnoreCase));
        photo = new XmppMucAvatarPhoto(
            ContentType: string.IsNullOrWhiteSpace(vCard.Photo.ContentType)
                ? XmppUserAvatar.RequiredContentType
                : vCard.Photo.ContentType,
            Data: bytes,
            Hash: hash,
            MatchedAdvertisedHash: matched);
        return true;
    }
}

public sealed record XmppMucAvatarPhoto(
    string ContentType,
    byte[] Data,
    string Hash,
    bool MatchedAdvertisedHash);

File: php/lib/Xmpp/XmppMucAvatar.php
<?php
declare(strict_types=1);

namespace Tiedragon\Xmpp;

final class XmppMucAvatar
{
    public const ROOMINFO_FORM_TYPE = 'http://jabber.org/protocol/muc#roominfo';
    public const AVATAR_HASH_FIELD = 'muc#roominfo_avatarhash';

    public static function serviceSupportsAvatars(string $discoInfoXml): bool
    {
        $info = XmppDisco::parseInfoResult($discoInfoXml);
        return in_array(XmppMuc::NS, $info['features'], true)
            && in_array(XmppXml::VCARD_TEMP_NS, $info['features'], true);
    }

    /**
     * @return list<string>
     */
    public static function parseRoomAvatarHashes(string $discoInfoXml): array
    {
        $forms = XmppDataForm::parseForms($discoInfoXml);
        foreach ($forms as $form) {
            $formType = $form['fields']['FORM_TYPE']['values'][0] ?? null;
            if ($formType !== self::ROOMINFO_FORM_TYPE) {
                continue;
            }

            $values = $form['fields'][self::AVATAR_HASH_FIELD]['values'] ?? [];
            return array_values(array_filter(
                array_map(static fn (string $value): string => strtolower(trim($value)), $values),
                static fn (string $value): bool => $value !== ''
            ));
        }

        return [];
    }

    public static function getRoomAvatarRequest(string $id, XmppJid|string $room): string
    {
        return XmppVCardTemp::getRequest($id, XmppJid::parse((string)$room)->bare());
    }

    public static function setRoomAvatarRequest(string $id, XmppJid|string $room, string $mediaType, string $imageBytes): string
    {
        return XmppVCardTemp::setRequest($id, [
            'PHOTO' => [
                'type' => $mediaType,
                'bytes' => $imageBytes,
            ],
        ], XmppJid::parse((string)$room)->bare());
    }

    public static function removeRoomAvatarRequest(string $id, XmppJid|string $room): string
    {
        return XmppVCardTemp::setRequest($id, [], XmppJid::parse((string)$room)->bare());
    }

    public static function avatarHash(string $imageBytes): string
    {
        return sha1($imageBytes);
    }

    /**
     * @param list<string> $advertisedHashes
     * @return array{type:string,bytes:string,hash:string,matched:bool}|null
     */
    public static function parseVerifiedAvatar(string $vcardXml, array $advertisedHashes): ?array
    {
        $vcard = XmppVCardTemp::parse($vcardXml);
        $photo = $vcard['PHOTO'];
        if ($photo === null || ($photo['bytes'] ?? null) === null) {
            return null;
        }

        $bytes = $photo['bytes'];
        $hash = self::avatarHash($bytes);
        $normalizedHashes = array_map(static fn (string $value): string => strtolower(trim($value)), $advertisedHashes);

        return [
            'type' => $photo['type'] ?: 'image/png',
            'bytes' => $bytes,
            'hash' => $hash,
            'matched' => $normalizedHashes === [] || in_array($hash, $normalizedHashes, true),
        ];
    }
}
